### PR TITLE
Persist busy state on hover

### PR DESCRIPTION
**Pull request by jameskoster**
This came up in https://github.com/WordPress/gutenberg/pull/59714.

## What?
Ensure 'busy' styling remains visible when hovering busy buttons.

## Why?
When clicking a button involves immediately invoking the busy state, you cannot currently see that the button is busy because the hover styles override the busy styles:

https://github.com/WordPress/gutenberg/assets/846565/1a986723-2605-4c57-b245-be38c90d0041

## How?
Do not apply hover styles when the button `is-busy`:

https://github.com/WordPress/gutenberg/assets/846565/13386325-ff9d-416b-b9a3-f1f602707046

## Testing Instructions
1. `npm run storybook:dev`
2. Navigate to the Button component
3. Toggle on the `busy` prop
4. Mouse over (or focus) the button preview and observe the animation persists
5. Check all button variants

## Sidenote
Should busy buttons still show the `pointer` cursor on hover? That seems a bit unexpected to me.

### CODE REVIEW

**Review by t-hamano — COMMENTED** [review]
If `isBusy` is `true`, how about setting `disabled` to `true` at the same time? This approach is already widely used and does not require changing the behavior of the `Button` component itself. The hover action will also be ignored.

```javascript
<Button
	isBusy={ isBusy }
	disabled={ isBusy }
>
	{ __( 'Delete' ) }
</Button>
```

**Comment by jameskoster** [comment]
So long as we're happy for busy buttons to be non-interactive that might work. Though, I suppose it would have to be `aria-disabled` so that the button can retain focus?

**Comment by t-hamano** [comment]
My understanding is that opting in to `__experimentalIsFocusable` is a good approach if you don't want to lose focus.

Not this:

```js
<Button
	aria-disabled={ isBusy }
	isBusy={ isBusy }
/>
```

like this:

```js
<Button
	disabled={ isBusy }
	isBusy={ isBusy }
	__experimentalIsFocusable
/>
```

Pinging @WordPress/gutenberg-components for confirmation just in case 🙏

**Comment by jameskoster** [comment]
There's may be a better way, but I pushed a commit which applies `aria-disabled` to busy buttons. Also fixed the active states.

**Comment by jameskoster** [comment]
@WordPress/gutenberg-components does someone have a moment to review this one? :)

File: packages/components/src/button/index.tsx [hunk 1]
```diff
@@ -174,7 +174,7 @@ export function UnforwardedButton(
 	const anchorProps: ComponentPropsWithoutRef< 'a' > =
 		Tag === 'a' ? { href, target } : {};
 
-	if ( disabled && isFocusable ) {
+	if ( ( disabled && isFocusable ) || isBusy ) {
```
Comment (tyxla) [inline packages/components/src/button/index.tsx:177]: I disagree with this change. IMHO you might want to keep the button appearing busy, but still allow it to be non-disabled. If you really want to disable it, we do have the `disabled` prop that allows you to do it.

File: packages/components/src/button/style.scss [hunk 2]
```diff
@@ -75,7 +75,7 @@
 		&:disabled:active:enabled,
 		&[aria-disabled="true"],
 		&[aria-disabled="true"]:enabled, // This catches a situation where a Button is aria-disabled, but not disabled.
-		&[aria-disabled="true"]:active:enabled {
+		&[aria-disabled="true"]:active:enabled:not(.is-busy) {
```
Comment (tyxla) [inline packages/components/src/button/style.scss:78]: This works as intended and no longer changes the hover state of busy buttons.

However, I'm not sure I agree with it. I think busy buttons should still have a hover state, and that hover state should just respect the busy button styles. 

Although most component libraries make loading/busy buttons disabled, I can definitely see a valid case for a busy button that's still not disabled. Maybe an optimistic "Add to Cart" button that remains enabled? Or any other counter-like button that counts something inline but also submits data to the server? We also have `__experimentalIsFocusable` so that fixes the a11y concerns about disabled busy buttons.

In simple terms, I'd expect if the button is not disabled but is busy, to have a hover state - darker/lighter color, but still appearing busy. Then if the busy button is also disabled, no hover style at all.

**Review by tyxla — COMMENTED** [review]
I'd rather keep `disabled` behavior be controlled by the `disabled` prop, and let `isBusy` control the appearance of the button. Is there a good reason to force busy buttons to also be disabled when we can do it with a separate prop?

File: packages/components/src/button/style.scss (hunk 2 above)
Comment (jameskoster) [reply packages/components/src/button/style.scss:78]: Fair point. I'm not quite grasping the examples, but I suppose allowing consumers to employ enabled & busy buttons is more flexible.

For the example in the OP I guess the solution would be to make the Button `busy` and `disabled`.

The styling of primary busy buttons might use a tweak, but that's one to look at separately.
